### PR TITLE
hooks: add eyed3.plugins and magic

### PR DESCRIPTION
**Pull request by zhu**
eyed3 package: eyed3.plugins search default plugins by searching .py files in plugins dir.
python-magic-bin package: collect datas and dynamic libs.

### CODE REVIEW

File: PyInstaller/hooks/hook-eyed3.plugins.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from PyInstaller.utils.hooks import collect_data_files, collect_submodules
+from os.path import basename, splitext
+
+plugins = filter(lambda data: basename(data[0])[0] not in ('_', '.') and data[0].endswith('.py'),
+        collect_data_files('eyed3.plugins', include_py_files=True))
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-eyed3.plugins.py]: Why not using  `collect_submodules`, which already collect submodules only?

File: PyInstaller/hooks/hook-eyed3.plugins.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+datas = plugins
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-eyed3.plugins.py]: Sure these should become `date` files and not hidden imports? The former will not be scanned for further dependencies.

File: PyInstaller/hooks/hook-eyed3.plugins.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+imports = list(set(map(lambda data: splitext(basename(data[0]))[0],
+        datas)))
+
+hiddenimports = imports
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-eyed3.plugins.py]: This looks like being wrong. Either add the plugins as submodules/hiddenimport *or* as data-files. Why is there a need to do both? Please recheck.

File: PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import eyed3.plugins
+from PyInstaller.utils.hooks import logger
+
+def pre_find_module_path(api):
+    api.search_dirs += eyed3.plugins.__path__
```
Comment (htgoebel) [inline PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py]: `pre_find_module_path` is *rarely* used. Sure this is required?

File: PyInstaller/hooks/hook-eyed3.plugins.py (hunk 1 above)
Comment (zhu) [reply PyInstaller/hooks/hook-eyed3.plugins.py]: I used the same process that the eyed3 collects its plugins.
eyed3 just collects '.py' files without initial '_' or '.'. `collect_submodule` may collect some useless modules that just has '.pyc' or '.pyo' but no '.py'.

File: PyInstaller/hooks/hook-eyed3.plugins.py (hunk 2 above)
Comment (zhu) [reply PyInstaller/hooks/hook-eyed3.plugins.py]: eyed3 searches '.py' files in the plugins directory and load them. So I have to add them to both datas and hidden imports.

File: PyInstaller/hooks/hook-eyed3.plugins.py (hunk 3 above)
Comment (zhu) [reply PyInstaller/hooks/hook-eyed3.plugins.py]: I found that PyInstaller cannot import data-files, or maybe I'm wrong.
It just worked when I add them both.
Is there any recipes?

File: PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py (hunk 4 above)
Comment (zhu) [reply PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py]: eyed3 imports thoses plugins like `import classic` but not `import eyed3.plugins.classic`. So I have to add the `plugins` directory to the python path.

**Comment by zhu** [comment]
The `magic` hook is in PR #4262 now.

**Comment by zhu** [comment]
https://github.com/nicfit/eyeD3/blob/bdfe08c7edd6c6b723f597383eba81ad8fbf9a39/src/eyed3/plugins/__init__.py#L60
Here's the code that eyed3 loads its plugins.

File: PyInstaller/hooks/hook-eyed3.plugins.py (hunk 3 above)
Comment (htgoebel) [reply PyInstaller/hooks/hook-eyed3.plugins.py]: Well, as these are data-file, they are not intended to be imported. You could tkeak your application to do so, but this is less useful as it requires every application developer to to the same.

There is no reciept to work around such a resticted implementation.

But again please use more readable code like this:
```
# TODO: remove `.py` extension
hiddenimports = ["eye3d.plugins." + os.path.basename(data[0])
    for data in datas]
```
No need to filter duplicates,

File: PyInstaller/hooks/hook-magic.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+from PyInstaller.utils.hooks import collect_data_files, collect_dynamic_libs
+
+datas = collect_data_files('magic')
+binaries = collect_dynamic_libs('magic')
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-magic.py]: Please remove this file from this pull-request.

File: PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py (hunk 4 above)
Comment (htgoebel) [reply PyInstaller/hooks/pre_find_module_path/hook-eyed3.plugins.py]: A `pre_find_module_path` hook only effects PyInstaller at freeze-time. But the hook already searches for the all sub-module in the correct directory. No need for this file.

File: news/4262.hooks.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Add hook for [eyed3.plugins](https://pypi.org/project/eyeD3/)
+Add hook for [magic](https://pypi.org/project/python-magic-bin)
```
Comment (htgoebel) [inline news/4262.hooks.rst]: The second line is too much.

**Comment by htgoebel** [comment]
You might want to have a look at a0683955e50354a9e70b4bcea87f6185ce156012 to get a notion of how helpful comments would look like.

**Comment by Legorooj** [comment]
Out of date; need moving to contrib soon anyway.

**Pull request closed without merging**